src/audio/eq_fir.c

```diff
@@ -575,28 +576,36 @@ static int fir_cmd_set_data(struct comp_dev *dev,
 			return -EBUSY;
 		}
 
-		/* Check and free old config */
-		eq_fir_free_parameters(&cd->config);
-
 		/* Copy new config, find size from header */
 		cfg = (struct sof_eq_fir_config *)cdata->data->data;
-		bs = cfg->size;
-		trace_eq("fir_cmd_set_data(): blob size: %u", bs);
+		bs = cdata->msg_id == 0 ? cfg->size : cdata->data->size;
+		trace_eq("fir_cmd_set_data(): blob size: %u msg_id %u", bs,
+			 cdata->msg_id);
 		if (bs > SOF_EQ_FIR_MAX_SIZE || bs == 0)
 			return -EINVAL;
 
-		/* Allocate buffer for copy of the blob. */
-		cd->config = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, bs);
-		if (!cd->config) {
-			trace_eq_error("fir_cmd_set_data() error: "
-				       "buffer allocation failed");
-			return -EINVAL;
+		if (cdata->msg_id == 0) {
+			/* Check and free old config */
```

Comment:
This makes me think we need a speculative rmalloc_is_avail() where we check if the new buffer size is free before we free the old coefficients. This way we guarantee that the EQ will still work with old coefficients if the new coefficients alloc fails.

Reply:
ok, I'll check it.

Reply:
Seems like we need an implementation of realloc(), but I would add a flag for the copy of old data as in most cases we wont need to copy.